src/arweave/composite-client.ts

```diff
@@ -554,62 +565,62 @@ export class ArweaveCompositeClient
     const url = `/${transactionType}/${txId}`;
     let downloadedFromPeer = true;
 
-    const responsePromise = this.txStore
-      .get(txId)
-      .then((tx) => {
+    const responsePromise = (async () => {
+      try {
+        // Check if it's already in the store
+        const storedTx = await this.txStore.get(txId);
+
         this.failureSimulator.maybeFail();
 
-        // Return cached TX if it exists
-        if (!this.skipCache && tx) {
-          return tx;
+        if (!this.skipCache && storedTx) {
+          return storedTx;
         }
 
-        return this.peerGetTx(url)
-          .catch(async () => {
-            downloadedFromPeer = false;
+        // Attempt to fetch from peer
+        let response;
 
-            // Request TX from trusted node if peer fetch failed
-            return this.trustedNodeRequestQueue.push({
-              method: 'GET',
-              url,
-            });
-          })
-          .then((response) => {
-            // Delete TX data to reduce response cache size
-            if (response?.data?.data) {
-              delete response.data.data;
-            }
-
-            return response.data;
-          });
-      })
-      .then(async (tx) => {
         try {
-          metrics.arweaveTxFetchCounter.inc({
-            node_type: downloadedFromPeer ? 'arweave_peer' : 'trusted',
+          response = await this.peerGetTx(url, 3);
+        } catch {
+          // If peer fails, fall back to trusted node
+          downloadedFromPeer = false;
+          response = await this.trustedNodeRequestQueue.push({
+            method: 'GET',
+            url,
           });
-          // Sanity check to guard against accidental bad data from both
-          // cache and trusted node
-          sanityCheckTx(tx);
-
-          await this.txStore.set(tx);
+        }
 
-          return tx;
-        } catch (error) {
-          this.txStore.del(txId);
+        // Delete the TX data payload (if present) to minimize memory/cache usage
+        if (response?.data?.data) {
+          delete response.data.data;
         }
-      })
-      .catch((error) => {
+
+        // Sanity-check the result
+        metrics.arweaveTxFetchCounter.inc({
+          node_type: downloadedFromPeer ? 'arweave_peer' : 'trusted',
+        });
+        sanityCheckTx(response.data);
+
+        // Store to our TX cache
+        await this.txStore.set(response.data);
+
+        return response.data;
+      } catch (error) {
+        // If something goes wrong, remove it from the store (in case partially cached)
+        this.txStore.del(txId);
         this.log.warn('Transaction prefetch failed:', {
-          txId: txId,
-          message: error.message,
-          stack: error.stack,
+          txId,
+          message: (error as Error).message,
```

Comment:
Nit: but can we push this `as` (or similar) up into the earlier `catch` line?

Reply:
in the catch clause itself: `Catch clause variable type annotation must be 'any' or 'unknown' if specified.ts(1196)` I'll reassign it to a new variable to cast it to Error.

Reply:
done